core/trino-main/src/main/java/io/trino/split/BufferingSplitSource.java

```diff
@@ -72,50 +78,71 @@ public Optional<List<Object>> getTableExecuteSplitsInfo()
     }
 
     private static class GetNextBatch
+            extends AbstractFuture<SplitBatch>
     {
         private final Context context = Context.current();
         private final SplitSource splitSource;
+        private final Executor executor;
         private final int min;
         private final int max;
 
+        @GuardedBy("this")
         private final List<Split> splits = new ArrayList<>();
-        private boolean noMoreSplits;
+        @GuardedBy("this")
+        private ListenableFuture<SplitBatch> nextBatchFuture;
 
         public static ListenableFuture<SplitBatch> fetchNextBatchAsync(
                 SplitSource splitSource,
+                Executor executor,
                 int min,
                 int max)
         {
-            GetNextBatch getNextBatch = new GetNextBatch(splitSource, min, max);
-            ListenableFuture<Void> future = getNextBatch.fetchSplits();
-            return Futures.transform(future, ignored -> new SplitBatch(getNextBatch.splits, getNextBatch.noMoreSplits), directExecutor());
+            GetNextBatch getNextBatch = new GetNextBatch(splitSource, executor, min, max);
+            getNextBatch.fetchSplits();
+            return getNextBatch;
         }
 
-        private GetNextBatch(SplitSource splitSource, int min, int max)
+        private GetNextBatch(SplitSource splitSource, Executor executor, int min, int max)
         {
             this.splitSource = requireNonNull(splitSource, "splitSource is null");
+            this.executor = requireNonNull(executor, "executor is null");
             checkArgument(min <= max, "Min splits greater than max splits");
             this.min = min;
             this.max = max;
         }
 
-        private ListenableFuture<Void> fetchSplits()
+        private synchronized void fetchSplits()
```

Comment:
Does it actually need to be synchronized? I think this is accessed by single thread at a time, no?

I think it's probably better to keep it though because it doesn't hurt

Reply:
It is accessed by single thread but it needs to be synchronised to guarantee the GuardedBy annotation on nextBatchFuture

Reply:
Yeah. I'm not sure `@GuardedBy` is needed (volatile is probably sufficient), but let's keep it as is.